code/modules/reagents/chemistry/recipes.dm

```diff
@@ -57,6 +78,101 @@
 	return
 	//I recommend you set the result amount to the total volume of all components.
 
+/**
+ * Stuff that occurs in the middle of a reaction
+ * Only procs DURING a reaction
+ * If reactionFlags & REACTION_INSTANT then this isn't called
+ *
+ * Arguments:
+ * * holder - the datum that holds this reagent, be it a beaker or anything else
+ * * created_volume - volume created per step
+ * * added_purity - how pure the created volume is per step
+ */
+/datum/chemical_reaction/proc/reaction_step(datum/reagents/holder, added_volume, added_purity)
+	return
+
+/**
+ * Stuff that occurs at the end of a reaction. This will proc if the beaker is forced to stop and start again (say for sudden temperature changes).
+ * Only procs at the END of reaction
+ * If reactionFlags & REACTION_INSTANT then this isn't called
+ * if reactionFlags REACTION_CLEAR_IMPURE then the impurity chem is handled here, producing the result in the beaker instead of in a mob
+ * Likewise for REACTION_CLEAR_INVERSE the inverse chem is produced at the end of the reaction in the beaker
+ * You should be calling ..() if you're writing a child function of this proc otherwise purity methods won't work correctly
+ *
+ * Proc where the additional magic happens.
+ * You dont want to handle mob spawning in this since there is a dedicated proc for that.client
+ * Arguments:
+ * * holder - the datum that holds this reagent, be it a beaker or anything else
+ * * react_volume - volume created across the whole reaction
+ */
+/datum/chemical_reaction/proc/reaction_finish(datum/reagents/holder, react_vol)
+	//failed_chem handler
+	for(var/id in results)
+		var/datum/reagent/R = holder.has_reagent(id)
+		if(!R)
+			continue
+		if(R.purity < purity_min)
+			var/cached_volume = R.volume
+			holder.remove_reagent(R.type, cached_volume, FALSE)
+			holder.add_reagent(R.failed_chem, cached_volume, FALSE, added_purity = 1)
+			SSblackbox.record_feedback("tally", "chemical_reaction", 1, "[type] failed reactions")
+			continue
+
+		//REACTION_CLEAR handler
+		if(reactionFlags == REACTION_CLEAR_IMPURE | REACTION_CLEAR_INVERSE)
+			if(R.purity == 1) //Failures can delete R
+				continue
+
+			var/cached_volume = R.volume
+			if(reactionFlags == REACTION_CLEAR_INVERSE && R.inverse_chem)
+				if(R.inverse_chem_val > R.purity)
+					holder.remove_reagent(R.type, cached_volume, FALSE)
+					holder.add_reagent(R.inverse_chem, cached_volume, FALSE, added_purity = 1)
+
+			else if (reactionFlags == REACTION_CLEAR_IMPURE && R.impure_chem)
```

Comment:
This logic won't work. I assume this is what you want:
```suggestion
		if(reactionFlags & (REACTION_CLEAR_IMPURE | REACTION_CLEAR_INVERSE))
			if(R.purity == 1) //Failures can delete R
				continue

			var/cached_volume = R.volume
			if(reactionFlags & REACTION_CLEAR_INVERSE && R.inverse_chem)
				if(R.inverse_chem_val > R.purity)
					holder.remove_reagent(R.type, cached_volume, FALSE)
					holder.add_reagent(R.inverse_chem, cached_volume, FALSE, added_purity = 1)

			else if (reactionFlags & REACTION_CLEAR_IMPURE && R.impure_chem)
```
Are `REACTION_CLEAR_INVERSE` and `REACTION_CLEAR_IMPURE` mutually exclusive?

Reply:
No, you can have a reagent with both flags if you want, but I see my error. I got rid of the else too. Is putting the brackets around the bitflag check being too cautious?

Reply:
I think bitwise & | and ^ in byond have same priority, so you have to put parentheses around all flags you check
for example:
`reactionFlags = 4 (100 binary)`
`flag1 = 1 (001 binary)`
`flag2 = 2 (010 binary)`
so if you write `reactionFlags & flag1 | flag2` you get `100 & 001 | 010 = 010` 
and if you write `reactionFlags & (flag1 | flag2)` you get `100 & (001 | 010) = 100 & 011 = 0`

Reply:
precedence is properly explained in the reference: <http://www.byond.com/docs/ref/index.html#/operator>

Reply:
> 
> 
> No, you can have a reagent with both flags if you want, but I see my error. I got rid of the else too. Is putting the brackets around the bitflag check being too cautious?

Using parentheses is fine, and more often than not add to clarity.

Reply:
This is very helpful to know, I've gotten very confused about these things before. Thank you for linking me this!